internal/app/machined/pkg/controllers/kubeaccess/crd.go

```diff
@@ -0,0 +1,204 @@
+			ID:        pointer.To(secrets.KubernetesID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: secrets.NamespaceName,
+			Type:      secrets.OSRootType,
+			ID:        pointer.To(secrets.OSRootID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *CRDController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *CRDController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	var crdControllerCtxCancel context.CancelFunc
+
+	crdControllerErrCh := make(chan error, 1)
+
+	stopCRDController := func() {
+		if crdControllerCtxCancel != nil {
+			crdControllerCtxCancel()
+
+			<-crdControllerErrCh
+
+			crdControllerCtxCancel = nil
+		}
+	}
+
+	defer stopCRDController()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil //nolint:govet
+		case <-r.EventCh():
+		case err := <-crdControllerErrCh:
+			if crdControllerCtxCancel != nil {
+				crdControllerCtxCancel()
+			}
+
+			crdControllerCtxCancel = nil
+
+			if err != nil && !errors.Is(err, context.Canceled) {
+				return fmt.Errorf("error from crd controller: %w", err)
+			}
+		}
+
+		kubeaccessConfig, err := safe.ReaderGet[*kubeaccess.Config](ctx, r, kubeaccess.NewConfig(config.NamespaceName, kubeaccess.ConfigID).Metadata())
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error fetching kubeaccess config: %w", err)
+			}
+
+			continue
+		}
+
+		var kubeaccessConfigSpec *kubeaccess.ConfigSpec
+
+		if kubeaccessConfig != nil {
+			kubeaccessConfigSpec = kubeaccessConfig.TypedSpec()
+		}
+
+		if kubeaccessConfig == nil || kubeaccessConfigSpec == nil || !kubeaccessConfigSpec.Enabled {
+			stopCRDController()
+
+			continue
+		}
+
+		kubeSecretsResources, err := safe.ReaderGet[*secrets.Kubernetes](ctx, r, resource.NewMetadata(
+			secrets.NamespaceName,
+			secrets.KubernetesType,
+			secrets.KubernetesID,
+			resource.VersionUndefined,
+		))
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error fetching kubernetes secrets: %w", err)
+			}
+
+			continue
+		}
+
+		kubeSecretsSpec := kubeSecretsResources.TypedSpec()
+
+		osSecretsResource, err := safe.ReaderGet[*secrets.OSRoot](ctx, r, resource.NewMetadata(
+			secrets.NamespaceName,
+			secrets.OSRootType,
+			secrets.OSRootID,
+			resource.VersionUndefined,
+		))
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error fetching os secrets: %w", err)
+			}
+
+			continue
+		}
+
+		osSecretsSpec := osSecretsResource.TypedSpec()
+
+		kubeconfig, err := clientcmd.BuildConfigFromKubeconfigGetter("", func() (*clientcmdapi.Config, error) {
+			return clientcmd.Load([]byte(kubeSecretsSpec.LocalhostAdminKubeconfig))
+		})
+		if err != nil {
+			return fmt.Errorf("error loading kubeconfig: %w", err)
+		}
+
+		stopCRDController()
```

Comment:
why is the controller being stopped here?

Reply:
Every time an event is received, we stop&start the controller so it is created with the new config (allowed roles & namespaces). Same idea as we do for kubelet: https://github.com/siderolabs/talos/blob/master/internal/app/machined/pkg/controllers/k8s/kubelet_service.go#L186-L190

An alternative would be to never stop the controller but pass the new variables in a concurrency-safe manner, but it didn't occur to me while writing it :)